src/operator/grid_generator-inl.h

```diff
@@ -0,0 +1,314 @@
+ public:
+  explicit GridGeneratorOp(GridGeneratorParam p) {
+    this->param_ = p;
+  }
+
+  virtual void Forward(const OpContext &ctx,
+                       const std::vector<TBlob> &in_data,
+                       const std::vector<OpReqType> &req,
+                       const std::vector<TBlob> &out_data,
+                       const std::vector<TBlob> &aux_args) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    CHECK_EQ(in_data.size(), 1);
+    CHECK_EQ(out_data.size(), 3);
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    switch (param_.transform_type) {
+      case grid::kAffine: {
+        // if transform_type is affine, data is affine matrix, input shape : (batch, 2, 3)
+        Tensor<xpu, 4, DType> out = out_data[grid::kOut].get<xpu, 4, DType>(s);
+        Tensor<xpu, 2, DType> grid_dst = out_data[grid::kGridDst].get<xpu, 2, DType>(s);
+        Shape<3> data_shape = Shape3(out.size(0), 2, 3);
+        Tensor<xpu, 3, DType> data = in_data[grid::kData]
+          .get_with_shape<xpu, 3, DType>(data_shape, s);
+        Tensor<cpu, 2, DType> workspace =
+              ctx.requested[grid::kTempSpace].get_host_space_typed<2, DType>(
+              grid_dst.shape_);
+        Tensor<xpu, 3, DType> tmp = out_data[grid::kTmp].get<xpu, 3, DType>(s);
+        for (index_t i = 1; i <= workspace.size(1); i++) {
+          // grid dst coordinate is (x, y, 1)
+          workspace[0][i-1] = -1.0 + (i-1) % param_.target_shape[1] * 2.0 /
+                              (param_.target_shape[1] - 1);
+          workspace[1][i-1] = -1.0 + (i-1) / param_.target_shape[1] * 2.0 /
+                              (param_.target_shape[0] - 1);
+          workspace[2][i-1] = 1.0;
+        }
+        // create a sampling grid, tmp : (batch, 2, H x W)
+        Copy(grid_dst, workspace, grid_dst.stream_);
+        for (index_t batch = 0; batch < data.size(0); batch++) {
+              tmp[batch] = dot(data[batch], grid_dst);
+        }
+        // out : (batch, 2, H, W)
+        out = reshape(tmp, out.shape_);
+        break;
+      }
+      // Warping transformation
+      case grid::kWarp: {
+        // if transform_type is warp, data is optical flow, input shape : (batch, 2, height, width)
+        // grid_src = grid_dst + optical flow
+        Tensor<xpu, 4, DType> data = in_data[grid::kData].get<xpu, 4, DType>(s);
+        Tensor<xpu, 4, DType> out = out_data[grid::kOut].get<xpu, 4, DType>(s);
+        // grid_dst : (2, H, W) similar to the results of numpy.meshgrid
+        Tensor<xpu, 3, DType> grid_dst = out_data[grid::kGridDst].get<xpu, 3, DType>(s);
+        Tensor<xpu, 1, DType> tmp = out_data[grid::kTmp].get<xpu, 1, DType>(s);
+        Tensor<xpu, 1, DType> workspace =
+              ctx.requested[grid::kTempSpace]
+                .get_space_typed<xpu, 1, DType>(Shape1(data.size(3)), s);
+        workspace = range<DType>(0, data.size(3));
+        grid_dst[0] = repmat(workspace, data.size(2));

```

Comment:
Could we use `grid_dst[0] = range<DType>(0, data.size(3), 1, data.size(2)).T()` instead?